polling_stations/apps/councils/management/commands/import_councils.py

```diff
@@ -276,40 +279,43 @@ def handle(self, **options):
-                CouncilGeography.objects.using(self.database).all().delete()
-            except UnsafeToDeleteCouncil:
-                self.stdout.write(
-                    "Can't delete councils there are polling stations attached. "
-                    "Please run manage.py teardown --all first if you want to delete them."
-                )
-                return
-
-        self.seen_ids = set()
-        self.import_councils_from_ec()
-
-        if not options["only_contact_details"]:
-            self.attach_boundaries(options.get("alt_url"))
-
-            if options.get("use_osni", None):
-                self.attach_boundaries(
-                    NI_BOUNDARY_URL, id_field=NI_ID_FIELD, include=NI_BOUNDARY_GSS_CODES
-                )
-
-            if options.get("import_old_boundaries", None):
-                self.attach_boundaries(
-                    OLD_BOUNDARY_URL,
-                    id_field=OLD_ID_FIELD,
-                    include=OLD_BOUNDARIES_GSS_CODES,
-                )
-
-        # Clean up old councils that we've not seen in the EC data
-        Council.objects.using(self.database).exclude(
-            council_id__in=self.seen_ids
-        ).delete()
-
-        self.stdout.write("..done")
+        with transaction.atomic(self.database):
+            if options["teardown"]:
+                try:
+                    self.stdout.write("Clearing councils table..")
+                    Council.objects.using(self.database).all().delete()
+                    self.stdout.write("Clearing councils_geography table..")
+                    CouncilGeography.objects.using(self.database).all().delete()
+                except UnsafeToDeleteCouncil:
+                    self.stdout.write(
+                        "Can't delete councils there are polling stations attached. "
+                        "Please run manage.py teardown --all first if you want to delete them."
+                    )
+                    return
+
+            self.seen_ids = set()
+            self.import_councils_from_ec()
+
+            if not options["only_contact_details"]:
+                self.attach_boundaries(options.get("alt_url"))
+
+                if options.get("use_osni", None):
+                    self.attach_boundaries(
+                        NI_BOUNDARY_URL,
+                        id_field=NI_ID_FIELD,
+                        include=NI_BOUNDARY_GSS_CODES,
+                    )
+
+                if options.get("import_old_boundaries", None):
+                    self.attach_boundaries(
+                        OLD_BOUNDARY_URL,
+                        id_field=OLD_ID_FIELD,
+                        include=OLD_BOUNDARIES_GSS_CODES,
+                    )
+
+            # Clean up old councils that we've not seen in the EC data
+            Council.objects.using(self.database).exclude(
+                council_id__in=self.seen_ids
+            ).delete()
+
+            self.stdout.write("..done")
```

Comment:
It looks like I changed a lot here because I've intented this whole function in one tab.
What I've actually done here is moved the `transaction.atomic` from being a decorator round this function to a `with` block so that I can reference `this.database`. One of the problems here was that the transaction was always on the principal DB regardless of the value passed in `--database`
You can look at the diff ignoring whitespace at https://github.com/DemocracyClub/UK-Polling-Stations/pull/8605/files?w=1 which is a bit more managable

Reply:
TiL `?w=1`. It also works on the commit diff view.